### ISSUE TITLE
Is there a way to postpone creating an instance?

### ISSUE THREAD

**Issue by xBodro** [simpleinjector/SimpleInjector#163]
Can I register whatever SomeType\<T\>(Func<T> resolver) so that resolving of T argument would happen when resolver is invoked? Good example is Lazy\<T\>.

**Comment by qujck** [comment]
Did you see the [docs](http://simpleinjector.readthedocs.org/en/latest/howto.html#register-factory-delegates)?

**Comment by xBodro** [comment]
Thank you for pointing that out.

Is it possible to do following thing:

container.Register<IMyService, MyService>();
container.Register<Lazy<>>(whatever here);

and then if I ask it for IMyService it would give me an instance of IMyService,
if I ask for Lazy<IMyService> it would give me Lazy<IMyService>

I just don't want to register every service as a generic argument of Lazy<> or as a LazyProxy as shown in documentation.

**Comment by dotnetjunkie** [comment]
Your question raises a red flag to me, because the creation of service components should be very quick, because [injection constructors should be simple and fast](http://blog.ploeh.dk/2011/03/03/InjectionConstructorsshouldbesimple/). The fact that you need `Lazy<T>` for multiple components is a strong indication that your components require heavy initialization.

The documentation even contains a warning about this:

> Warning: The same warning applies to the use of `Lazy<T>` as it does for the use of `Func<T>` delegates. Further more, the constructors of your components should be simple, reliable and quick, and that would remove the need for lazy initialization. For more information about creating an application and container configuration that can be successfully verified, please read the How To Verify the container’s configuration.

Can you describe why you think you need to inject `Lazy<T>`s?

**Comment by xBodro** [comment]
Yeah, such code seems to smell at a glance. There are no other reasons for that than to save another millisecond or a byte of memory.
There are cases when deep object graph requires multiple objects to be created (i.e. hundreds of instances), however execution may stop at the very top of the graph, making all nested objects useless. Isn't it overkill to create  whole graph of objects if we can be satisfied by Lazy, that would trigger further evaluation of graph if needed?

**Comment by dotnetjunkie** [comment]
Hi @xBodro, compared to some other DI libraries, auto-resolving `Lazy<T>` is not supported out-of-the-box in Simple Injector. This behavior can be plugged in quite easily though, as you can see in the [ResolvingFactoriesExtensions](https://github.com/simpleinjector/SimpleInjector/blob/master/SimpleInjector.CodeSamples/ResolvingFactoriesExtensions.cs) code sample. But there's a very good reason that Simple Injector doesn't support this out-of-the-box by Simple Injector.

First of all, Simple Injector is blazingly fast and unless you do too much in your constructors, there should typically not often be a reason to do this in Simple Injector. And please be aware of premature optimizations. Do benchmark this before starting to do this.

But second, even if you decide that you need to delay building a part of the object graph, you should prevent injecting `Lazy<T>` throughout your application code. The injection of `Lazy<T>` into your application code is a [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) violation, because you leak implementation details of the component into the consumer. You leak implementation details, because now the consumer of the `Lazy<T>` becomes aware about the fact that the component is heavy to create. And it not only becomes aware of this, but it immediately complicates the consumer's code, because it has to deal with `Lazy<T>` and we complicate the consumer's unit tests, because they have to inject an `Lazy<T>` into the consumer under test.

So even if your application requires object graphs to be built lazily, you only want to break the object graph at certain specific places, and you don't want to leak this complexity into consumers (and you don't want to change consumers because of this). Instead, you want to wrap such component into a proxy implementation that is able to postpone the creation of the object graph. This might still mean that the proxy implemention depends on `Lazy<T>` in its constructor. For instance:

``` c#
public interface ISomeComponent {
    void SomeMethod();
}

public class LazySomeComponentProxy : ISomeComponent {
    private readonly Lazy<ISomeComponent> lazy;
    public LazySomeComponentProxy(Lazy<ISomeComponent> lazy) { this.lazy = lazy; }
    public void SomeMethod() { this.lazy.Value.SomeMethod(); }
}
```
This proxy implementation allows the rest of the application oblivious about the fact the the object graph is delayed. Registration can be done as follows:

``` c#
container.Register<RealComponent>();
container.Register<ISomeComponent>(() => new LazySomeComponentProxy(
    new Lazy<ISomeComponent>(container.GetInstance<RealComponent>)));
```

I hope this helps.

**Comment by xBodro** [comment]
Thank you very much.
